contracts/cw1-whitelist-ng/src/contract.rs

```diff
@@ -0,0 +1,440 @@
+use cosmwasm_std::{
+    from_slice, Addr, Api, Binary, CosmosMsg, Deps, DepsMut, Env, MessageInfo, Response, StdError,
+    StdResult,
+};
+use serde::de::DeserializeOwned;
+
+use crate::error::ContractError;
+use crate::interfaces::*;
+use crate::msg::{
+    AdminListResponse, Cw1ExecMsg, Cw1QueryMsg, InstantiateMsg, WhitelistExecMsg, WhitelistQueryMsg,
+};
+use crate::state::{AdminList, Cw1WhitelistContract};
+
+use cw1::CanExecuteResponse;
+use cw2::set_contract_version;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:cw1-whitelist";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+pub fn validate_admins(api: &dyn Api, admins: &[String]) -> StdResult<Vec<Addr>> {
+    admins.iter().map(|addr| api.addr_validate(&addr)).collect()
+}
+
+impl<T> Cw1WhitelistContract<T> {
+    pub fn instantiate(
+        &self,
+        deps: DepsMut,
+        _env: Env,
+        _info: MessageInfo,
+        admins: Vec<String>,
+        mutable: bool,
+    ) -> Result<Response<T>, ContractError> {
+        set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+        let cfg = AdminList {
+            admins: validate_admins(deps.api, &admins)?,
+            mutable,
+        };
+
+        self.admin_list.save(deps.storage, &cfg)?;
+        Ok(Response::new())
+    }
+
+    pub fn is_admin(&self, deps: Deps, addr: &str) -> Result<bool, ContractError> {
+        let cfg = self.admin_list.load(deps.storage)?;
+        Ok(cfg.is_admin(addr))
+    }
+
+    // Entry points, to be called only in actual entry points and by multitest `Contract`
+    // implementation
+    pub(crate) fn entry_instantiate(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        msg: &[u8],
+    ) -> Result<Response<T>, ContractError> {
+        let msg: InstantiateMsg = from_slice(msg)?;
+        msg.dispatch(deps, env, info, self)
+    }
+
+    pub(crate) fn entry_execute(
+        &self,
+        deps: DepsMut,
+        env: Env,
+        info: MessageInfo,
+        msg: &[u8],
+    ) -> Result<Response<T>, ContractError>
+    where
+        T: DeserializeOwned,
+    {
+        let mut errs = vec![];
+
+        match from_slice::<Cw1ExecMsg<T>>(msg) {
```

Comment:
This multiple attempts to deserialize is a nice approach. It looks a little ugly here, but I see the beauty of each message being able to implement dispatch and the contract being able to implement multiple traits.

If this is auto-generated (which it will be shortly), then I definitely see the bonus.